src/relations/database_provides.py

```diff
@@ -1,142 +1,70 @@
-        )
-
-        self.charm.app_peer_data[CREDENTIALS_SHARED] = "true"
-
-    def _on_database_broken(self, _) -> None:
-        """Handle the database relation broken event."""
-        self.charm.unit.status = WaitingStatus(
-            f"Waiting for relations: {DATABASE_PROVIDES_RELATION}"
-        )
-        if not self.charm.unit.is_leader():
-            return
-
-        # application user cleanup when backend relation still in place
-        if backend_relation := self.charm.model.get_relation(DATABASE_REQUIRES_RELATION):
-            if app_data := self.charm.app_peer_data.get(MYSQL_ROUTER_REQUIRES_APPLICATION_DATA):
-                username = json.loads(app_data)["username"]
-
-                db_username = backend_relation.data[backend_relation.app]["username"]
-                db_password = backend_relation.data[backend_relation.app]["password"]
-                db_host, db_port = backend_relation.data[backend_relation.app]["endpoints"].split(
-                    ":"
-                )
-
-                MySQLRouter.delete_application_user(
-                    username=username,
-                    hostname="%",
-                    db_username=db_username,
-                    db_password=db_password,
-                    db_host=db_host,
-                    db_port=db_port,
-                )
-        # clean up departing app data
-        self.charm.app_peer_data.pop(MYSQL_ROUTER_REQUIRES_APPLICATION_DATA, None)
-        self.charm.app_peer_data.pop(MYSQL_ROUTER_PROVIDES_DATA, None)
-        self.charm.app_peer_data.pop(CREDENTIALS_SHARED, None)
-        self.charm.set_secret("app", "application-password", None)
+import dataclasses
+import secrets
+import string
+
+import charms.data_platform_libs.v0.data_interfaces as data_interfaces
+import ops
+
+from constants import PASSWORD_LENGTH
+
+
+@dataclasses.dataclass
+class Relation:
+    interface: data_interfaces.DatabaseProvides
+
+    @property
+    def active(self) -> bool:
+        """Whether relation is currently active"""
+        for key in ["database", "username", "password", "endpoints"]:
+            if key not in self._local_databag:
+                return False
+        return True
+
+    @property
+    def database(self) -> str:
+        return self._remote_databag["database"]
+
+    @property
+    def username(self) -> str:
+        return f"relation-{self._id}"
+
+    @property
+    def _relation(self) -> ops.model.Relation:
+        relations = self.interface.relations
+        assert len(relations) == 1
+        return relations[0]
+
+    @property
+    def _id(self) -> int:
+        return self._relation.id
+
+    @property
+    def _remote_databag(self) -> dict:
+        return self.interface.fetch_relation_data()[self._id]
+
+    @property
+    def _local_databag(self) -> ops.model.RelationDataContent:
+        return self._relation.data[self.interface.local_app]
+
+    def is_desired_active(self, event) -> bool:
```

Comment:
could be just me, but the method naming is not very telling.
```suggestion
    def is_event_relation_active(self, event) -> bool:
```